pkg/zonemgr/zone_file.go

```diff
@@ -2,180 +2,151 @@ package zonemgr
+	if zoneFile.nameServerName != "" {
+		zoneFile.headerSuf += fmt.Sprintf("IN NS %s.\n", zoneFile.nameServerName)
 	}
 
-	if zoneFile.userNameserverIP != "" {
-		zoneFile.headerSuf += fmt.Sprintf("IN A %s\n", zoneFile.userNameserverIP)
+	if zoneFile.nameServerIP != "" {
+		zoneFile.headerSuf += fmt.Sprintf("IN A %s\n", zoneFile.nameServerIP)
 	}
 }
 
 func (zoneFile *ZoneFile) generateHeader() (header string) {
 	return zoneFile.headerPref + strconv.Itoa(zoneFile.soaSerial) + zoneFile.headerSuf
 }
 
-//////////// maps impl
-
-func (zoneFile *ZoneFile) updateVMIRecords(vmiName string, vmiNamespace string, interfaces []v1.VirtualMachineInstanceNetworkInterface) (isUpdated bool, err error) {
+func (zoneFile *ZoneFile) updateVMIRecords(vmiName string, vmiNamespace string, interfaces []v1.VirtualMachineInstanceNetworkInterface) (isUpdated bool) {
 	key := fmt.Sprintf("%s_%s", vmiName, vmiNamespace)
 	isUpdated = false
 	//delete vmi records from the list
 	if interfaces == nil {
-		delete(zoneFile.vmiRecordsMap, key)
-		isUpdated = true
-	} else {
-		ifacesMap := zoneFile.vmiRecordsMap[key]
-		//if not exist or changed - override
-		if ifacesMap == nil || isIfaceListChanged(ifacesMap, interfaces) {
-			zoneFile.vmiRecordsMap[key] = buildInterfacesMap(vmiName, vmiNamespace, interfaces)
+		if zoneFile.vmiRecordsMap[key] != nil {
+			delete(zoneFile.vmiRecordsMap, key)
 			isUpdated = true
 		}
+	} else {
+		newRecords := buildARecordsArr(vmiName, vmiNamespace, interfaces)
+		isUpdated = !isEqual(newRecords, zoneFile.vmiRecordsMap[key])
+		if isUpdated {
+			zoneFile.vmiRecordsMap[key] = newRecords
+		}
 	}
 
-	fmt.Println("********** zoneFile.vmiRecordsMap: ", zoneFile.vmiRecordsMap)
-
 	if isUpdated {
 		zoneFile.updateContent()
 	}
-	return isUpdated, nil
-}
-
-type ARecord struct {
-	value string
-	ip    string
+	return isUpdated
 }
 
-func isIfaceListChanged(ifacesMap map[string]ARecord, interfaces []v1.VirtualMachineInstanceNetworkInterface) (isChanged bool) {
-	if len(ifacesMap) != len(interfaces) {
-		return true
+func isEqual(arr1, arr2 []string) bool {
+	if len(arr1) != len(arr2) {
+		return false
 	}
-
-	for _, newIface := range interfaces {
-		if ifacesMap[newIface.Name].ip != newIface.IP {
-			return true
+	sort.Strings(arr1)
+	sort.Strings(arr2)
+	for i, v := range arr1 {
+		if v != arr2[i] {
+			return false
 		}
 	}
-	return false
+	return true
 }
 
-func buildInterfacesMap(vmiName string, vmiNamespace string, interfaces []v1.VirtualMachineInstanceNetworkInterface) (ifaceMap map[string]ARecord) {
-	ifaceMap = make(map[string]ARecord)
+func buildARecordsArr(vmiName string, vmiNamespace string, interfaces []v1.VirtualMachineInstanceNetworkInterface) (recordsArr []string) {
 	for _, iface := range interfaces {
-		ifaceMap[iface.Name] = ARecord{value: generateARecord(vmiName, vmiNamespace, iface.Name, iface.IP), ip: iface.IP}
+		if iface.Name != "" {
+			IPs := iface.IPs
+			for _, IP := range IPs {
+				if net.IsIPv4String(IP) {
```

Comment:
can an interface have multi IPv4 ?
maybe yes, but don't remember we support it
if not you can break / return upon the first find 
btw even if we do support it on kubevirt
we can't support here a FQDN because the same FQDN will have two ips

Reply:
added `break`